src/VisualStudio/Core/Def/RoslynPackage.cs

```diff
@@ -144,10 +144,30 @@ protected override void OnSaveOptions(string key, Stream stream)
 
     protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
     {
+        // Should only be called from a threadpool. Opinionated, as package load sequence thread switches are impactful.
+        Contract.ThrowIfTrue(JoinableTaskFactory.Context.IsOnMainThread);
```

Comment:
Ditt.

Reply:
Not sure what the scenario is where this is called on the UI thread, but I've switched to now ensure we're off it when starting our work